### ISSUE TITLE
Do not enforce client cert and key for `ClientOptions::with_ssl`

### ISSUE THREAD

**Issue by lukaspustina** [mongodb-labs/mongo-rust-driver-prototype#235]
Dear crate team,

thanks for this helpful crate for which I'd like to propose a change.

The function `ClientOptions::with_ssl` enforces a crate user to supply a CA cert as well as a client cert and a corresponding key. This is unfortunate, because this requirement steams from the code and not from the actual mongo DB driver. For example, it's perfectly fine to create an SSL encrypted connection using only use a CA cert -- cf. https://docs.mongodb.com/manual/tutorial/configure-ssl-clients/, Connect to MongoDB Instance with SSL Encryption.

The requirement is encoded in https://github.com/mongodb-labs/mongo-rust-driver-prototype/blob/master/src/stream.rs#L81. It'd be very easy to mitigate this requirement by using `Option` for at least `certificate_file` and `key_file` and set the openssl ssl_context accordingly. Unfortunately, introducing `Option` would break the current API of `ClientOptions::with_ssl`. The other way around this is to check for the empty string `""` which is a anti-pattern.

Anyway, probably you guys have a better pictures of this and I'd like to start a discussion to evolve this. I'm happy to contribute code, once we agree on a design.

Cheers,
Lukas

**Comment by saghm** [comment]
I think you might be misreading the documentation that you linked to. Specifically, the [`--sslPEMKeyFile`](https://docs.mongodb.com/manual/reference/program/mongostat/#cmdoption-sslpemkeyfile) option that the documentation describes requires that a client certificate is present in the file as well. This driver slightly diverges from that API by accepting the client certficate and client key as separate files instead of as a single file (although we plan to add the ability to use a single file as well), but the actual requirements are the same for the client credentials. (The requirements as a whole are slightly different in that the driver currently requires specifying a custom CA certificate  rather than defaulting to using the system CA store, but I don't think this is what you're asking about).

**Comment by lukaspustina** [comment]
Thanks for your quick response. I think I didn't make myself clear.

The issue I like to resolve is that the current implementation requires to supply a CA certificate as well as client certificate. For a transport encrypted connection the client certificate is not necessary, but has to be supplied anyway though.

I'd like to change the crate in a way that only the CA certificate has to be passed to the factory method enabling TLS _without_ a client certificate. Without suggesting this change, but for demonstration purpose, a modified method signature could be:

```
pub fn with_ssl(ca_file: &str, client_cert_and_key: Option<(&str, &str)>, verify_peer: bool) -> Self
```

**Comment by saghm** [comment]
Ah, okay, I think I understand now! Supporting encryption without authentication definitely should be supported here. I don't think the change would need to be very involved; the only slight challenge would be deciding the most ergonomic way to allow users to either provide both parts of the client credential or neither.

**Comment by lukaspustina** [comment]
I completely agree. The main question is whether to try to keep the current API working (minor version) change or introduce an API breaking change (major version).

From the ergonomic point of view, we should change the API and move the client certificate parameters into some kind of Option. I'll be happy to prepare a PR for this. Just give me the go.

It's up to you then to decide how to create a new version for this.

**Comment by saghm** [comment]
Feel free to submit a PR! I'm not opposed to changing the API slightly for this if necessary; we still consider the API to be unstable (hence the pre-1.0 version). That being said, I'm not sure that having an `Option<(&str, &str)>` is the most ergonomic way to do this; my instinct is to have two separate methods, once that takes the four existing parameters and one that just takes the CA certificate and the verify peer flag. Naming these two methods would be the trickier part here; maybe calling the new function with only two arguments `with_unauthenticated_ssl` would be suitable (and non-API breaking, since then we could leave the other method as-is), and we could add document that the user will need to use `Database::auth` if they want to authenticate their connection made with this method (which might currently have to be done anyhow;  I haven't had a chance to test X509 auth with the driver yet).

**Comment by lukaspustina** [comment]
cf. https://github.com/mongodb-labs/mongo-rust-driver-prototype/pull/238 please.

**Comment by saghm** [comment]
Merged in #238 and published in 0.3.5. Thanks again for your help with this!
